### DISCUSSION TITLE
Academic Research on TypeSpec for API-First Design

### DISCUSSION THREAD

**Discussion by NielsCo** [microsoft/typespec#1941] [General]
Hey everyone,

### Background
I'm currently deep into a thesis project where I'm using TypeSpec to build a web app, specifically creating stubs for it, following the API-First Design approach.

With TypeSpec, I can generate OpenAPI Specifications (OAS), which are then used by tools like the OpenAPI Generator to create stubs for both the frontend and backend. However, I noticed there weren't many tools out there that create database schemas from OAS.

To fill this gap, and to keep all the relevant info inside the TypeSpec (so I don't have to manually edit the OAS), I've developed a TypeSpec-Postgres emitter. I'm currently talking with my company and university about making this open-source.

###  Question
Despite a lot of searching, I haven't found any academic research that specifically mentions TypeSpec (or its previous name, CADL). Has anyone come across any research on this?

### Outlook
API-First Design is still pretty new in academia, as highlighted in the 2022 paper "API-First Design: A Survey of the State of Academia and Industry" by Nicole Beaulieu, Sergiu M. Dascalu, and Emily Hand. I'm expecting new research on the topic to pop up during my thesis work.

TypeSpec is a great tool for describing cloud service APIs and generating other API description languages, client and service code, documentation, and more. My work on the TypeSpec-Postgres emitter further shows the potential of TypeSpec in maintaining the integrity of the API-First Design approach. I think academic research on this tool could really contribute to the field of API-First Design.

If you have any insights or resources, I'd love to hear about them. Thanks for your time!

**Comment by daviwil** [comment]
Hi @NielsCo, there isn't any existing research that I'm aware of, but @bterlson would certainly be interested to hear about your thesis project!

**Reply by NielsCo** [reply]
Hey David. Thanks for your answer. Brian has my contact info if he has any more questions about the [prototype](https://github.com/NielsCo/typespec-postgres) (postgres-emitter) 😊
The thesis itself will sadly not be published as it contains some confidential parts. 
I haven't looked into other methods of publishing the parts of the thesis that can be made public. But I certainly would be interested in recommendations.
